Gems/Atom/RHI/Code/Include/Atom/RHI/MultiDevicePipelineLibrary.h

```diff
@@ -0,0 +1,130 @@
+#include <Atom/RHI/MultiDeviceObject.h>
+#include <Atom/RHI/PipelineLibrary.h>
+
+#include <Atom/RHI.Reflect/PipelineLibraryData.h>
+
+namespace AZ::RHI
+{
+    //! A handle typed to the pipeline library. Used by the PipelineStateCache to abstract access.
+    using MultiDevicePipelineLibraryHandle = Handle<uint32_t, class MultiDevicePipelineLibrary>;
+
+    //! A descriptor struct containing a map of device-specific PipelineLibraryDescriptors, used in
+    //! MultiDevicePipelineLibrary
+    struct MultiDevicePipelineLibraryDescriptor
+    {
+        //! Returns the device-specific PipelineLibraryDescriptor for the given index
+        inline PipelineLibraryDescriptor GetDevicePipelineLibraryDescriptor(int deviceIndex) const
+        {
+            AZ_Assert(
+                m_devicePipelineLibraryDescriptors.find(deviceIndex) != m_devicePipelineLibraryDescriptors.end(),
+                "No DevicePipelineLibraryDescriptor found for device index %d\n",
+                deviceIndex);
+
+            if (m_devicePipelineLibraryDescriptors.contains(deviceIndex))
+            {
+                return m_devicePipelineLibraryDescriptors.at(deviceIndex);
+            }
+            else
+            {
+                return PipelineLibraryDescriptor{};
+            }
+        }
+
+        //! A map of all device-specific PipelineLibraryDescriptors, indexed by the device index
+        AZStd::unordered_map<int, PipelineLibraryDescriptor> m_devicePipelineLibraryDescriptors;
+    };
+
+    //! MultiDevicePipelineState initialization is an expensive operation on certain platforms. If multiple pipeline states
+    //! are created with little variation between them, the contents are still duplicated. This class is an allocation
+    //! context for pipeline states, provided at MultiDevicePipelineState::Init, which will perform de-duplication of
+    //! internal pipeline state components and cache the results.
+    //!
+    //! Practically speaking, if many pipeline states are created with shared data between them (e.g. permutations
+    //! of the same shader), then providing a MultiDevicePipelineLibrary instance will reduce the memory footprint and cost
+    //! of compilation.
+    //!
+    //! Additionally, the MultiDevicePipelineLibrary is able to serialize the internal driver-contents to and from an opaque
+    //! data blob. This enables building up a pipeline state cache on disk, which can dramatically reduce pipeline
+    //! state compilation cost when run from a pre-warmed cache.
+    //!
+    //! MultiDevicePipelineLibrary is thread-safe, in the sense that it will take a lock during compilation. It is possible
+    //! to initialize pipeline states across threads using the same MultiDevicePipelineLibrary instance, but this will
+    //! result in the two calls serializing on the mutex. Instead, see PipelineStateCache which stores
+    //! a MultiDevicePipelineLibrary instance per thread to avoid this contention.
```

Comment:
This is a copy paste of PipelineLibrary and really MultiDevicePipelineLibrary doesn't have code to do this work. Let's re-do all the comments here to match what the class is doing which is mostly to do with managing SingleDevice pipeline libraries and possibly merging them.